File: src/main/java/me/mrkirby153/KirBot/command/CommandExecutor.java
package me.mrkirby153.KirBot.command;

import me.mrkirby153.KirBot.user.Clearance;

public abstract class CommandExecutor {

    private String[] aliases;

    private String description;

    private Clearance clearance;

    public abstract void execute();

    public String[] getAliases() {
        return aliases;
    }

    public void setAliases(String[] aliases) {
        this.aliases = aliases;
    }

    public Clearance getClearance() {
        return clearance;
    }

    public void setClearance(Clearance clearance) {
        this.clearance = clearance;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }
}

File: src/main/kotlin/me/mrkirby153/KirBot/command/executors/CommandExecutor.kt
package me.mrkirby153.KirBot.command.executors

import me.mrkirby153.KirBot.Shard
import me.mrkirby153.KirBot.data.ServerData
import me.mrkirby153.KirBot.user.Clearance
import net.dv8tion.jda.core.Permission
import net.dv8tion.jda.core.entities.Guild
import net.dv8tion.jda.core.entities.Message
import net.dv8tion.jda.core.entities.User

abstract class CommandExecutor {

    lateinit var command: String

    var aliases: Array<String> = arrayOf<String>()

    var description: String = "No descrpition provided"

    var clearance: Clearance = Clearance.USER

    var permissions: Array<Permission> = arrayOf()

    var category: String = "Uncategoriezed"

    lateinit var shard: Shard

    lateinit var serverData: ServerData

    lateinit var guild: Guild

    abstract fun execute(message: Message, args: Array<String>)

    protected fun getUserByMention(mention: String): User? {
        val id = mention.replace("[<@!>]".toRegex(), "")

        return shard.getUserById(id)
    }
}